beginner_source/profiler_tutorial.py

```diff
@@ -0,0 +1,205 @@
+"""
+Profiling your PyTorch Module
+------------
+
+PyTorch includes a profiler API that is useful to identify the time and
+memory costs of various PyTorch operations in your code. It records the
+events of functions executed in C++, and exposes them to Python.
+Profiler can be easily integrated in your code, and the results are
+printed in a pretty table or as a JSON trace file.
+
+.. note::
+    Profiler supports multithreaded models. Profiler runs in the
+    same thread as the operation but it will also profile child operators
+    that might run in another thread. Concurrently-running profilers will be
+    scoped to their own thread to prevent mixing of results.
+
+Head on over to `this
+recipe <https://pytorch.org/tutorials/recipes/recipes/profiler.html>`__
+for a quick walkthrough of the Profiler API.
+
+"""
+
+import torch
+import numpy as np
+from torch import nn
+import torchvision.models as models
+import torch.autograd.profiler as profiler
+
+
+######################################################################
+# Performance debugging using Profiler
+# ~~~~~~~~~~~~~~~~~
+#
+# Profiler can be useful to identify performance bottlenecks in your
+# models. In this example, we build a custom module that performs two
+# sub-tasks:
+#
+# - a linear transformation on the input, and
+# - use the transformation result to get indices on a mask tensor.
+#
+# We wrap the code for each sub-task in separate labelled context managers using
+# ``profiler.record_function("label")``. In the profiler output, the
+# aggregate performance metrics of all operations in the sub-task will
+# show up under its corresponding label.
+#
+#
+# Note that this incurs some overhead, and is best used only while
+# investigating the trace. Remove it if you are benchmarking runtimes.
+#
+
+class CustomLinear(torch.nn.Linear):
+    def __init__(self, in_features: int, out_features: int, bias: bool = True) -> None:
+        super().__init__(in_features, out_features, bias)
+
+    def forward(self, input, mask):
+        # with profiler.record_function("LINEAR PASS"):
+            out = super().forward(input)
+
+        # with profiler.record_function("MASK INDICES"):
+            threshold = out.sum(axis=1).mean().item()
+            hi_idx = np.argwhere(mask.cpu().numpy() > threshold) # fix: use torch.nonzero()
```

Comment:
should we move it back to gpu?

Reply:
i think it's not necessary because moving to cpu alone is pretty expensive and showcases inefficiencies of using numpy where torch functions could have worked